src/mlpack/methods/ann/vanilla_rbm.hpp

```diff
@@ -0,0 +1,207 @@
+  * @param i Index of point to use for objective function evaluation.
+  * 
+  */
+  double Evaluate(const arma::mat& parameters, const size_t i);
+
+ /** 
+  * This function calculates
+  * the free energy of the model
+  * @param: input data point 
+  */
+  double FreeEnergy(arma::mat&& input);
+
+ /*
+  * This functions samples the hidden
+  * layer given the visible layer
+  *
+  * @param input visible layer input
+  * @param output the sampled hidden layer
+  */
+  void SampleHidden(arma::mat&& input, arma::mat&& output);
+
+  /*
+  * This functions samples the visble
+  * layer given the hidden layer
+  *
+  * @param input hidden layer
+  * @param output the sampled visible layer 
+  */
+  void SampleVisible(arma::mat&& input, arma::mat&& output);
+
+ /*
+  * This function does the k-step
+  * gibbs sampling.
+  *
+  * @param input: input to the gibbs function
+  * @param output: stores the negative sample
+  */
+  void Gibbs(arma::mat input, arma::mat&& output);
+
+  /*
+   * Calculates the gradients for the rbm network
+   *
+   * @param input index the visible layer/data point
+   * @param neg_input stores the negative samples computed usign the gibbs 
+   * @param k number of steps for gibbs sampling
+   * @param persistence pcdk / not 
+   * @param output store the gradients
+   */
+  void Gradient(const size_t input, arma::mat& output);
+
+  //! Return the number of separable functions (the number of predictor points).
+  size_t NumFunctions() const { return numFunctions; }
+
+  //! Return the initial point for the optimization.
+  const arma::mat& Parameters() const { return parameter; }
+  //! Modify the initial point for the optimization.
+  arma::mat& Parameters() { return parameter; }
+
+  VisibleLayerType& VisibleLayer() { return visible; }
+  HiddenLayerType& HiddenLayer() { return hidden; }
+
+  //! Serialize the model.
+  template<typename Archive>
+  void Serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  /* 
+   * ForwardVisible layer compute the forward
+   * activations given the visible layer
+   *
+   * @param input the visible layer
+   * @param output the acitvation function
+   */
+  void ForwardVisible(arma::mat&& input, arma::mat&& output)
+  {
+    visible.Forward(std::move(input), std::move(output));
+  };
+
+  /* 
+   * ForwardHidden layer compute the forward
+   * activations given the hidden layer
+   *
+   * @param input the visible layer
+   * @param output the acitvation function
+   */
+  void ForwardHidden(arma::mat&& input, arma::mat&& output)
+  {
+    hidden.Forward(std::move(input), std::move(output));
+  };
+
+  /*
+   * Helper function for Gradient
+   * calculates the gradients for both
+   * positive and negative samples.
+   */
+  void CalcGradient(arma::mat&& input, arma::mat&& output)
+  {
+    arma::mat temp1;
+    ForwardVisible(std::move(input), std::move(temp1));
+    output = temp1 * input.t();
+    output.reshape(output.n_rows * output.n_cols, 1);
```

Comment:
As for me, this code is unsafe since it assumes that armadilo stores matrices in the column major order.
Could you rewrite that using `arma::mat(something.memptr(), ....)`?